jetty-util/src/main/java/org/eclipse/jetty/util/BufferUtil.java

```diff
@@ -971,11 +971,19 @@ public static boolean isMappedBuffer(ByteBuffer buffer)
 
     public static ByteBuffer toBuffer(Resource resource, boolean direct) throws IOException
     {
-        int len = (int)resource.length();
+        long len = resource.length();
         if (len < 0)
             throw new IllegalArgumentException("invalid resource: " + resource + " len=" + len);
 
-        ByteBuffer buffer = direct ? BufferUtil.allocateDirect(len) : BufferUtil.allocate(len);
+        if (len > Integer.MAX_VALUE)
+        {
+            // This method cannot handle resources of this size.
+            return null;
```

Comment:
Are callers of this method prepared to receive `null`?
Before this change, this method was never returning `null`.
I would rather throw if it's an exceptional case.

Reply:
I also would like to see a test for this case (either if we return `null` or throw). It is already covered?

Reply:
Yes, there are only 4 callers for this `BufferUtil.toBuffer(Resource, boolean)` method.
Excluding test cases, there's only the 2 flavors of `.getDirectBuffer()` and `.getIndirectBuffer()` that use this method.
All 4 usages of `.getDirectBuffer()` and `.getIndirectBuffer()` handle null and expect it.

Reply:
Throwing causes an exception to be logged at WARN level every time.
This isn't necessary based on how `.getDirectBuffer()` and `.getIndirectBuffer()` are used.
If those are unavailable, (by returning null), then the `Resource.getReadableByteChannel()` or `Resource.getInputStream()` are used instead.

See HttpOutput.sendContent() for example usage:

https://github.com/eclipse/jetty.project/blob/4ebf4b7477acb42c73a069a37e29b5fd0738bfec/jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java#L987-L1020

Reply:
After conversation with @gregw we added another safety check for buffers before the `BufferUtil.toBuffer(Resource, boolean)` usage in the CachedContentFactory.